### PR TITLE
HIP Update

### PR DESCRIPTION
**Pull request by rchen20**
Updates CoMD, Lulesh 1.0, and 2.0 to use the HIP backend. Requires two other changes before merging:

1. Integrate latest CHAI into Kripke (https://github.com/LLNL/Kripke/pull/35).
2. Update BLT/RAJA/Camp into Proxies (https://github.com/LLNL/RAJAProxies/pull/27).
3. Requires HIP updates in Kripke (https://github.com/LLNL/Kripke/pull/42).
4. Update HIP.  <== This PR.

### CODE REVIEW

**Comment by rchen20** [comment]
@davidbeckingsale I'm getting this Umpire/Camp HIP error, even though it seems that all of the `ENABLE_HIP`s are being set. Have you seen this before?

[ 56%] Building CXX object kripke/tpl/umpire/src/umpire/op/CMakeFiles/umpire_op.dir/HipCopyOperation.cpp.o
cd /usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/build_lc_toss4-amdclang-4.5.2-gfx908/kripke/tpl/umpire/src/umpire/op && /opt/rocm-4.5.2/llvm/bin/amdclang++  -D__HIP_PLATFORM_AMD__=1 -D__HIP_PLATFORM_HCC__=1 -I/usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/tpl/RAJA/tpl/camp/include -I/usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/build_lc_toss4-amdclang-4.5.2-gfx908/tpl/RAJA/tpl/camp/include -I/usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/kripke/tpl/umpire/src -I/usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/build_lc_toss4-amdclang-4.5.2-gfx908/include -isystem /opt/rocm-4.5.2/include  -Wall -Wextra      -O2 -fPIC   -std=c++14 -o CMakeFiles/umpire_op.dir/HipCopyOperation.cpp.o -c /usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/kripke/tpl/umpire/src/umpire/op/HipCopyOperation.cpp
/usr/workspace/wsrzc/chen59/allraja/rajaproxyupdate/rajaproxy_git_proxyhip/kripke/tpl/umpire/src/umpire/op/HipCopyOperation.cpp:33:42: error: no member named 'Hip' in namespace 'camp::resources'
  auto device = ctx.get<camp::resources::Hip>();

**Comment by davidbeckingsale** [comment]
Hey @rchen20 I've not seen that before - I would double check that the camp, blt, umpire etc. versions are all pointing at the correct commit.

You might be missing a CAMP_ENABLE_HIP, we have some code in Umpire to do that here: https://github.com/LLNL/Umpire/blob/develop/src/tpl/CMakeLists.txt#L110

It might be nicer/easier to have a standalone camp subdir in Kripke rather than pull in one from RAJA.

**Comment by rchen20** [comment]
Hello @rhornung67 @davidbeckingsale, this is ready to go (after another Kripke update with HIP fixes). Please take a look when you have a chance.

**Comment by rchen20** [comment]
Friendly reminder: this is ready for review, please take a look!

File: CoMD/src-threaded/CMakeLists.txt [hunk 1]
```diff
@@ -73,6 +77,18 @@ target_compile_options(
   coMD_threaded
   PUBLIC
   ${comd_compile_options})
+elseif(ENABLE_HIP)
+  include_directories(
```
Comment (davidbeckingsale) [inline CoMD/src-threaded/CMakeLists.txt:81]: Are these all really required? If so, they should be `target_include_directories` specific to the targets that require them.

**Review by davidbeckingsale — COMMENTED** [review]
Just one quick question.

File: CoMD/src-threaded/CMakeLists.txt (hunk 1 above)
Comment (rchen20) [reply CoMD/src-threaded/CMakeLists.txt:81]: Thanks, I was duplicating this logic from the CUDA and MPI versions of this. I'll look in to making this a bit better. It's odd that the `CXXFLAGS` are added.

File: CoMD/src-threaded/CoMD.cpp [hunk 2]
```diff
@@ -207,6 +209,16 @@ SimFlat* initSimulation(Command cmd)
...
+#elif defined(ENABLE_HIP)
+   /* This is a hint to the HIP runtime that this structure is mostly read only.  This
+    * creates read-only copies on the CPU and GPU to avoid unnecessary thrashing as this
+    * structure is needed throughout the code.  Technically, this structure should only
+    * be accessed (dereferenced) on the GPU once GPU code starts but this is apparently
+    * insufficient to avoid this thrashing.
+    */
+   //int device;
+   //hipGetDevice(&device);
+   // No hipMemAdvise available.
```
[the remark this answers is not in the document]
Comment (MrBurmark) [reply CoMD/src-threaded/CoMD.cpp:221]: But there's no reason to not put this in device memory in hip as that is accessible from the host.

[the remark this answers is not in the document]
Comment (rchen20) [reply CoMD/src-threaded/CoMD.cpp]: Yes, CoMD is using `hipMalloc` anyway, so this should be unnecessary.

**Comment by rchen20** [comment]
@davidbeckingsale @MrBurmark Took out some unnecessary stuff, and fixed HIP and CUDA malloc's in CoMD. Please take another look!

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
Thanks @rchen20 After merging this, please tag a release and indicate what's been updated in the release.

**Pull request merged**